### ISSUE TITLE
java.lang.ClassNotFoundException: org.postgresql.pljava.example.annotation.ComplexTuple

### ISSUE THREAD

**Issue by ChristophBerg** [tada/pljava#178]
On some of the distributions on apt.postgresql.org, installing the examples jar in pljava aborts with errors:

Seen on stretch/amd64, with PG 9.6 (9.3/4/5 worked):
```
20:07:05 + psql -c 'CREATE EXTENSION pljava'
20:07:05 CREATE EXTENSION
20:07:05 + psql -c 'SELECT sqlj.install_jar('\''file:/usr/share/postgresql/9.6/pljava/pljava-examples-1.5.1-BETA2.jar'\'',
'\''examples'\'', true)'
[...]
20:07:07 INFO:  16 Sep 18 18:07:07 org.postgresql.pljava.management.Commands Type id = OID(16686)
20:07:07 ERROR:  java.lang.ClassNotFoundException: org.postgresql.pljava.example.annotation.ComplexTuple
20:07:07 CONTEXT:  SQL statement "SELECT javatest.assertHasValues( CAST('(1,2)' AS javatest.complextuple), 1, 2)"
```

Seen on bionic/amd64, with 9.4 (9.3 worked):
```
20:06:37 + psql -c 'SELECT sqlj.install_jar('\''file:/usr/share/postgresql/9.4/pljava/pljava-examples-1.5.1-BETA2.jar'\'',
'\''examples'\'', true)'
[...]
20:06:38 ERROR:  java.lang.ClassNotFoundException: org.postgresql.pljava.example.annotation.Point
20:06:38 CONTEXT:  SQL statement "SELECT javatest.assertHasValues(CAST('(1,2)' AS point), 1, 2)"
```

**Comment by jcflack** [comment]
I don't think this is really distribution-specific; rather, if you re-run the the builds on all your distributions, you'll find that which ones fail keeps shifting around like Igor's hump in _Young Frankenstein_.

What's happening is that the order of commands in the generated deployment descriptor for the 'examples' jar is not fully deterministic. It is constrained by provides/requires relationships, but otherwise comes out in hashtable-influenced order. Every so often, it will generate an order where the use of a user-defined type precedes any other action that loads that class, and then the classloading fails because PostgreSQL assumes the type I/O functions don't need an updated snapshot. (And they don't, _provided the class is already loaded_, but if it hasn't been, the ClassLoader ends up working from a snapshot taken before the jar was installed, so of course it doesn't find the class). This shows how the `read_only` parameter to SPI functions can have surprising effects beyond just preventing writes.

This had been handled in the past by adding no-op non-immutable functions to affected classes, and arranging for those to be called early. But that's a hack and it is easy to overlook cases. I will work on a fix that doesn't require that.

Meanwhile, you really should find that a build that fails this way will usually miraculously be fixed if you just build it again.

**Comment by ChristophBerg** [comment]
From the perspective of https://reproducible-builds.org/, it would be nice if the build artifact was entirely deterministic... Debian likes that very much :)

**Comment by jcflack** [comment]
That should perhaps be opened as a separate feature request, as it would involve adding some arbitrary tie-breaking sort step to the existing topological order, serving no purpose other than a condition on reproducibility. But it could be made optional for just such occasions.

On reflection, I like the idea of an option. Non-determinism better exercises the possibilities (without it, this issue could have gone longer undetected), but I see the appeal of reproducibility also.

-Chap

**Comment by jcflack** [comment]
Gewiss. I'm not opposed to implementing it. But it ought to have its own issue, as it will be slightly non-trivial, and I'd like to use this issue for correcting this particular bug that it happened to flush out.

**Comment by jcflack** [comment]
In case I left too much unsaid: this is not simply a matter of a static ordering for a fixed set of examples built along with PL/Java: PL/Java eats its own dogfood and uses its SQL generator to produce the example DDR. The required change is in the SQL generator itself, which will also affect users compiling their own jars.

The cost/benefit trade will be the same for them: if their build works, it will work every time. If they then change/add/remove something in their code and the next build fails, it will fail every time, but not necessarily because of any problem in what was just changed/added/removed, but one that may have been latent for much longer and masked by the deterministic build.

Now, the most expected cause for an ordering that doesn't work would be that they overlooked some `provides`/`requires` declaration that was logically necessary, and the obvious thing to do is fix that. But in the case of an issue like this one, which is more subtle and really not the fault of the code being built, it could be challenging to know what to change.

Without the forced determinism, if the user was in a hurry to produce a jar that didn't fail, a simple repetition of the build command would probably do it. With forced determinism, the problem would be a blocker until the underlying cause could be found and fixed (or until the user changed/added some other random code to shake up the forced arbitrary order).

All that being said, it is possible that this issue involving Java classloading could be the only member of that "subtle" category, and fixing it may leave only user-overlooked `provides`/`requires` as reasons for a built jar to fail to install.
